File: leetcode/python/0345_reverse_vowels_of_a_string.py
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        if len(s) == 1: return s

        sList = list(s)
        l, r = 0, len(s) - 1
        while l < r:
            if not sList[l] in "aeiouAEIOU": l += 1
            if not sList[r] in "aeiouAEIOU": r -= 1
            if sList[l] in "aeiouAEIOU" and sList[r] in "aeiouAEIOU": 
                sList[l], sList[r] = sList[r], sList[l]
                l += 1
                r -= 1
        return "".join(sList)

File: leetcode/cpp/0345_reverse_vowels_of_a_string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        set<char> vowels={'a','e','i','o','u','A','E','I','O','U'};
        stack<char> stack;
        for(char c:s)
        {
            if(vowels.count(c)) stack.push(c);
        }
        for(char& c:s)
        {
            if(vowels.count(c)) 
            {
                c=stack.top();
                stack.pop();
            }
        }
        return s;
    }
};